src/forwarder-vppagent/main.go

```diff
@@ -0,0 +1,117 @@
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+)
+
+const (
+	// NSMEnvVariablePrefix - all viper keys are transformed to all caps and prefixed with "%s_" % NSMEnvVariablePrefix
+	NSMEnvVariablePrefix = "NSM"
+	// EndpointNameKey - name of the endpoint for use in NSM
+	EndpointNameKey = "endpoint_name"
+	// TunnelIPKey - IP we can use for originating and terminating tunnels
+	TunnelIPKey = "tunnel_ip"
+	// VppagentURLKey - URL for reaching VPPAgent
+	VppagentURLKey = "vppagent_url"
+	// BasedirKey - key for the BaseDir to use for the memif mechanism
+	BasedirKey = "base_dir"
+	// ListenOnURLKey - Viper key for retrieving the URL to listen for incoming networkservicemesh RPC calls
+	ListenOnURLKey = "listen_on_url"
+	// ConnectToURKey - Viper key for retrieving the URL to make an outgoing networkservicemesh RPC calls
+	ConnectToURKey = "connect_to_url"
+)
+
+func main() {
+	// Capture signals to cleanup before exiting - note: this *must* be the first thing in main
+	c := make(chan os.Signal, 1)
+	signal.Notify(c,
+		os.Interrupt,
+		// More Linux signals here
+		syscall.SIGHUP,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+
+	// Context to use for all things started in main
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	// Use Viper to capture information from the environment
+	viper.SetEnvPrefix(NSMEnvVariablePrefix)
+	_ = viper.BindEnv(TunnelIPKey)
+	_ = viper.BindEnv(BasedirKey)
+	_ = viper.BindEnv(ListenOnURLKey)
+	viper.AutomaticEnv()
+
+	// grpc.ClientConn for dialing the vppagent
+	vppagentCC, _ := grpc.DialContext(ctx, viper.GetString(VppagentURLKey))
+
+	// Tunnel IP for use for VXLAN and other Tunnel types
+	tunnelIP := net.ParseIP(viper.GetString(TunnelIPKey))
+
+	connectToURL, err := url.Parse(viper.GetString(ConnectToURKey))
+	if err != nil {
+		logrus.Fatalf("Please specify valid URL in %s Env Variable.  \"%s\" is not a valid URL", ConnectToURKey, viper.GetString(ConnectToURKey))
+	}
+
+	// XConnect Network Service Endpoint
+	endpoint := xconnectns.NewServer(
+		viper.GetString(EndpointNameKey),
+		vppagentCC,
+		viper.GetString(BasedirKey),
+		tunnelIP,
+		connectToURL, // url to reach up
+	)
+
+	// URL to listen on
+	listenOnURL, err := url.Parse(viper.GetString(ListenOnURLKey))
+	if err != nil {
+		logrus.Fatalf("Please specify valid URL in %s Env Variable.  \"%s\" is not a valid URL", ListenOnURLKey, viper.Get(ListenOnURLKey))
+	}
+
+	// Create listener
+	ln, err := net.Listen(listenOnURL.Scheme, listenOnURL.Host)
+	if err != nil {
+		logrus.Fatalf("failed to listen: %v", err)
+	}
+
+	// Create GRPC Server
+	// TODO - add ServerOptions for Tracing, Security, etc
+	server := grpc.NewServer()
+	networkservice.RegisterNetworkServiceServer(server, endpoint)
+
+	// Serve
+	err = server.Serve(ln)
+	if err != nil {
+		logrus.Fatalf("failed to listen on %s", listenOnURL)
+	}
+
+	// Wait for signals
+	<-c
```

Comment:
it will never start waiting for signals, because server above will Serve() indefinitely

Reply:
Yep... figured that out (in the code I'm still poking at :) ).

Reply:
Fixed